open-apparel-registry/open-apparel-registry
# Superusers Can View All Lists

## Overview

Adds a list view in the dashboard for Superusers that allows them to see a read-only view of any list any contributor has uploaded. The view is very similar to that of My Lists, with the inclusion of the Date Created column. Upon clicking through to a list, they are presented with a "read-only" view of the list, that allows for searching and filtering as usual, but does not allow confirming or rejecting potential matches.

Connects #547 

## Demo

![image](https://user-images.githubusercontent.com/1430060/59391939-e7099180-8d43-11e9-8f98-a50393a34fa0.png)

![2019-06-12 18 21 43](https://user-images.githubusercontent.com/1430060/59390811-5d0bf980-8d40-11e9-9857-6fe4b8b41755.gif)

## Notes

To facilitate the front-end, a few fields were exposed in the back-end that were previously not included: `is_superuser` for the user, `created_at` and `contributor_id` for the lists. If there are any security implications for exposing these fields, we should consider and address them now.

Much of the front-end was copied from other parts of the app, leading to a lot of duplication and redundancy in terms of styling and table markup. This is a good candidate for refactoring and abstracting, that can be done in a future pass.

There is an edge case when a superuser accesses their own list from the Dashboard Lists page. In this case, when they click "Back to Lists", they will be taken to their My Lists page instead. This is unlikely to be a huge issue immediately, since most Superusers don't have their own lists, but should be kept in mind. This can be addressed by changing "Back to Lists" from a link to simulate the browser's "Back" button, but I don't know if React Router provides such facility. It could also be implemented by setting a flag somewhere (either in the URL or in Redux), but I did not deem it important enough to include in this pass. Please let me know if I should.

## Testing Instructions

- Check out this branch and `server`
- Go to [:6543](http://localhost:6543/) and sign in as a Superuser
  - A Superuser is any user that has `is_superuser=True` in the database
- Go to [:6543/dashboard/lists/](http://localhost:6543/dashboard/lists/)
  - [ ] Ensure you see a table and a select
- Open the select box
  - [ ] Ensure it is populated with a list of contributors
- Select a contributor
  - [ ] Ensure the contributor's lists are fetched from the server and populate the table
  - [ ] Ensure the lists are sorted newest first
  - [ ] Ensure the URL is updated with the selected contributor
- Click a list
  - [ ] Ensure you are taken to that list's detail page
- Observe that the view is "read-only"
  - [ ] Ensure searching / filtering works as usual
  - [ ] Ensure you cannot confirm or reject potential matches
- Click "Back to lists"
  - [ ] Ensure you are taken to Dashboard Lists, rather than your My Lists page
  - [ ] Ensure Dashboard Lists has the correct contributor selected in the view and the URL
- Refresh the page
  - [ ] Ensure the contributor is picked up from the URL and used to populate the select and the table
- Select a contributor that hasn't uploaded any lists
  - [ ] Ensure you see an empty message

## Checklist
